File: crates/ork-sdk-rust/src/deferrables.rs
//! Deferrable types for long-running external jobs
//!
//! Deferrables allow tasks to start long-running external jobs (BigQuery, Cloud Run, etc.)
//! and return immediately. The Ork scheduler tracks these jobs via the Triggerer component,
//! allowing workers to scale to zero while jobs run.
//!
//! # Example
//!
//! ```rust,ignore
//! use ork_sdk_rust::deferrables::BigQueryJob;
//!
//! fn my_task(input: MyInput) -> BigQueryJob {
//!     // Start BigQuery job
//!     let job = start_bigquery_job();
//!
//!     // Return deferrable - worker can scale to 0
//!     BigQueryJob {
//!         project: "my-project".to_string(),
//!         job_id: job.id,
//!         location: "US".to_string(),
//!     }
//! }
//! ```

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Trait for deferrable job types
///
/// Deferrables represent long-running external jobs that the scheduler
/// should track. Implementing this trait allows the scheduler to:
/// 1. Detect that a task returned a deferrable
/// 2. Extract job tracking information
/// 3. Route to the appropriate JobTracker implementation
pub trait Deferrable: Serialize + Send + Sync {
    /// Get the service type for routing to the correct JobTracker
    fn service_type(&self) -> &str;

    /// Get job identifier for tracking
    fn job_id(&self) -> String;

    /// Serialize deferrable to JSON for database storage
    fn to_json(&self) -> serde_json::Value {
        serde_json::to_value(self).unwrap_or(serde_json::Value::Null)
    }
}

/// BigQuery job deferrable
///
/// Used for long-running BigQuery queries. The scheduler will poll
/// the BigQuery API until the job completes.
///
/// # Example
///
/// ```rust,ignore
/// use google_cloud_bigquery::client::Client;
/// use ork_sdk_rust::deferrables::BigQueryJob;
///
/// async fn run_analytics() -> BigQueryJob {
///     let client = Client::new("my-project").await.unwrap();
///     let job = client.query("SELECT * FROM dataset.table").await.unwrap();
///
///     BigQueryJob {
///         project: "my-project".to_string(),
///         job_id: job.job_id().to_string(),
///         location: "US".to_string(),
///     }
/// }
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BigQueryJob {
    /// GCP project ID
    pub project: String,
    /// BigQuery job ID
    pub job_id: String,
    /// BigQuery location (e.g., "US", "EU")
    pub location: String,
}

impl Deferrable for BigQueryJob {
    fn service_type(&self) -> &str {
        "bigquery"
    }

    fn job_id(&self) -> String {
        format!("{}/{}/{}", self.project, self.location, self.job_id)
    }
}

/// Cloud Run job deferrable
///
/// Used for Cloud Run jobs. The scheduler will poll the Cloud Run API
/// until the job execution completes.
///
/// # Example
///
/// ```rust,ignore
/// use ork_sdk_rust::deferrables::CloudRunJob;
///
/// async fn trigger_batch_job() -> CloudRunJob {
///     let execution_id = start_cloud_run_job("my-job").await;
///
///     CloudRunJob {
///         project: "my-project".to_string(),
///         region: "us-central1".to_string(),
///         job_name: "my-job".to_string(),
///         execution_id,
///     }
/// }
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CloudRunJob {
    /// GCP project ID
    pub project: String,
    /// Cloud Run region
    pub region: String,
    /// Cloud Run job name
    pub job_name: String,
    /// Execution ID
    pub execution_id: String,
}

impl Deferrable for CloudRunJob {
    fn service_type(&self) -> &str {
        "cloudrun"
    }

    fn job_id(&self) -> String {
        format!(
            "{}/{}/{}/{}",
            self.project, self.region, self.job_name, self.execution_id
        )
    }
}

/// Dataproc job deferrable
///
/// Used for Dataproc Spark/Hadoop jobs. The scheduler will poll the
/// Dataproc API until the job completes.
///
/// # Example
///
/// ```rust,ignore
/// use ork_sdk_rust::deferrables::DataprocJob;
///
/// async fn run_spark_job() -> DataprocJob {
///     let job_id = submit_dataproc_job("my-cluster", "spark-job.py").await;
///
///     DataprocJob {
///         project: "my-project".to_string(),
///         region: "us-central1".to_string(),
///         cluster_name: "my-cluster".to_string(),
///         job_id,
///     }
/// }
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataprocJob {
    /// GCP project ID
    pub project: String,
    /// Dataproc region
    pub region: String,
    /// Cluster name
    pub cluster_name: String,
    /// Job ID
    pub job_id: String,
}

impl Deferrable for DataprocJob {
    fn service_type(&self) -> &str {
        "dataproc"
    }

    fn job_id(&self) -> String {
        format!(
            "{}/{}/{}/{}",
            self.project, self.region, self.cluster_name, self.job_id
        )
    }
}

/// Custom HTTP polling deferrable
///
/// Used for custom external services that provide HTTP status endpoints.
/// The scheduler will poll the URL until the response indicates completion.
///
/// # Example
///
/// ```rust,ignore
/// use ork_sdk_rust::deferrables::CustomHttp;
/// use std::collections::HashMap;
///
/// async fn trigger_external_job() -> CustomHttp {
///     let job_id = start_external_job().await;
///
///     let mut headers = HashMap::new();
///     headers.insert("Authorization".to_string(), "Bearer token".to_string());
///
///     CustomHttp {
///         url: format!("https://api.example.com/jobs/{}/status", job_id),
///         method: "GET".to_string(),
///         headers,
///         success_status_codes: vec![200],
///         completion_field: "status".to_string(),
///         completion_value: "completed".to_string(),
///         failure_value: Some("failed".to_string()),
///     }
/// }
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CustomHttp {
    /// URL to poll for job status
    pub url: String,
    /// HTTP method (GET, POST, etc.)
    pub method: String,
    /// Optional HTTP headers
    #[serde(default)]
    pub headers: HashMap<String, String>,
    /// HTTP status codes that indicate successful polling (default: [200])
    #[serde(default = "default_success_codes")]
    pub success_status_codes: Vec<u16>,
    /// JSON field to check for completion status
    pub completion_field: String,
    /// Value indicating job completion
    pub completion_value: String,
    /// Optional value indicating job failure
    pub failure_value: Option<String>,
}

fn default_success_codes() -> Vec<u16> {
    vec![200]
}

impl Deferrable for CustomHttp {
    fn service_type(&self) -> &str {
        "custom_http"
    }

    fn job_id(&self) -> String {
        // Use URL as job identifier
        self.url.clone()
    }
}

/// Deferrable wrapper for task outputs
///
/// Tasks can return either:
/// - Regular data (synchronous completion)
/// - One or more deferrables (asynchronous tracking)
/// - Both regular data and deferrables
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum TaskResult<T> {
    /// Synchronous result with regular data
    Data(T),
    /// Asynchronous result with deferrables
    Deferred {
        /// List of deferrables to track
        deferred: Vec<serde_json::Value>,
    },
    /// Mixed result with both data and deferrables
    Mixed {
        /// Regular output data
        data: T,
        /// List of deferrables to track
        deferred: Vec<serde_json::Value>,
    },
}

impl<T> TaskResult<T> {
    /// Create a result with only data
    pub fn data(value: T) -> Self {
        TaskResult::Data(value)
    }

    /// Create a result with only deferrables
    pub fn deferred(deferrables: Vec<Box<dyn Deferrable>>) -> Self {
        let deferred = deferrables
            .iter()
            .map(|d| d.to_json())
            .collect();
        TaskResult::Deferred { deferred }
    }

    /// Create a result with both data and deferrables
    pub fn mixed(data: T, deferrables: Vec<Box<dyn Deferrable>>) -> Self {
        let deferred = deferrables
            .iter()
            .map(|d| d.to_json())
            .collect();
        TaskResult::Mixed { data, deferred }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bigquery_job_service_type() {
        let job = BigQueryJob {
            project: "test-project".to_string(),
            job_id: "job123".to_string(),
            location: "US".to_string(),
        };
        assert_eq!(job.service_type(), "bigquery");
        assert_eq!(job.job_id(), "test-project/US/job123");
    }

    #[test]
    fn test_cloudrun_job_service_type() {
        let job = CloudRunJob {
            project: "test-project".to_string(),
            region: "us-central1".to_string(),
            job_name: "my-job".to_string(),
            execution_id: "exec123".to_string(),
        };
        assert_eq!(job.service_type(), "cloudrun");
        assert_eq!(job.job_id(), "test-project/us-central1/my-job/exec123");
    }

    #[test]
    fn test_dataproc_job_service_type() {
        let job = DataprocJob {
            project: "test-project".to_string(),
            region: "us-central1".to_string(),
            cluster_name: "my-cluster".to_string(),
            job_id: "job123".to_string(),
        };
        assert_eq!(job.service_type(), "dataproc");
        assert_eq!(job.job_id(), "test-project/us-central1/my-cluster/job123");
    }

    #[test]
    fn test_custom_http_service_type() {
        let job = CustomHttp {
            url: "https://api.example.com/status".to_string(),
            method: "GET".to_string(),
            headers: HashMap::new(),
            success_status_codes: vec![200],
            completion_field: "status".to_string(),
            completion_value: "done".to_string(),
            failure_value: Some("failed".to_string()),
        };
        assert_eq!(job.service_type(), "custom_http");
        assert_eq!(job.job_id(), "https://api.example.com/status");
    }

    #[test]
    fn test_task_result_serialization() {
        let result: TaskResult<String> = TaskResult::data("hello".to_string());
        let json = serde_json::to_string(&result).unwrap();
        assert!(json.contains("hello"));
    }
}

File: python-sdk/ork_sdk/deferrables.py
"""Deferrable types for long-running external jobs

Deferrables allow tasks to start long-running external jobs (BigQuery, Cloud Run, etc.)
and return immediately. The Ork scheduler tracks these jobs via the Triggerer component,
allowing workers to scale to zero while jobs run.

Example:
    from ork_sdk import BigQueryJob
    from google.cloud import bigquery

    def my_task(input_data):
        # Start BigQuery job
        client = bigquery.Client(project="my-project")
        job = client.query("SELECT * FROM dataset.table")

        # Return deferrable - worker can scale to 0
        return BigQueryJob(
            project="my-project",
            job_id=job.job_id,
            location="US"
        )
"""

from dataclasses import dataclass, field, asdict
from typing import Dict, List, Optional, Any, Union
from abc import ABC, abstractmethod
import json


class Deferrable(ABC):
    """Base class for deferrable job types

    Deferrables represent long-running external jobs that the scheduler
    should track. Implementing this class allows the scheduler to:
    1. Detect that a task returned a deferrable
    2. Extract job tracking information
    3. Route to the appropriate JobTracker implementation
    """

    @abstractmethod
    def service_type(self) -> str:
        """Get the service type for routing to the correct JobTracker"""
        pass

    @abstractmethod
    def tracking_id(self) -> str:
        """Get a stable identifier for tracking this deferred job"""
        pass

    def to_dict(self) -> Dict[str, Any]:
        """Serialize deferrable to dictionary for JSON storage"""
        payload = asdict(self)
        payload["service_type"] = self.service_type()
        payload.setdefault("job_id", self.tracking_id())
        return payload

    def to_json(self) -> str:
        """Serialize deferrable to JSON string"""
        return json.dumps(self.to_dict())


@dataclass
class BigQueryJob(Deferrable):
    """BigQuery job deferrable

    Used for long-running BigQuery queries. The scheduler will poll
    the BigQuery API until the job completes.

    Example:
        from google.cloud import bigquery
        from ork_sdk import BigQueryJob

        def run_analytics():
            client = bigquery.Client(project="my-project")
            job = client.query("SELECT * FROM dataset.table")

            return BigQueryJob(
                project="my-project",
                job_id=job.job_id,
                location="US"
            )

    Attributes:
        project: GCP project ID
        job_id: BigQuery job ID
        location: BigQuery location (e.g., "US", "EU")
    """

    project: str
    job_id: str
    location: str

    def service_type(self) -> str:
        return "bigquery"

    def tracking_id(self) -> str:
        return f"{self.project}/{self.location}/{self.job_id}"


@dataclass
class CloudRunJob(Deferrable):
    """Cloud Run job deferrable

    Used for Cloud Run jobs. The scheduler will poll the Cloud Run API
    until the job execution completes.

    Example:
        from ork_sdk import CloudRunJob

        def trigger_batch_job():
            execution_id = start_cloud_run_job("my-job")

            return CloudRunJob(
                project="my-project",
                region="us-central1",
                job_name="my-job",
                execution_id=execution_id
            )

    Attributes:
        project: GCP project ID
        region: Cloud Run region
        job_name: Cloud Run job name
        execution_id: Execution ID
    """

    project: str
    region: str
    job_name: str
    execution_id: str

    def service_type(self) -> str:
        return "cloudrun"

    def tracking_id(self) -> str:
        return f"{self.project}/{self.region}/{self.job_name}/{self.execution_id}"


@dataclass
class DataprocJob(Deferrable):
    """Dataproc job deferrable

    Used for Dataproc Spark/Hadoop jobs. The scheduler will poll the
    Dataproc API until the job completes.

    Example:
        from ork_sdk import DataprocJob

        def run_spark_job():
            job_id = submit_dataproc_job("my-cluster", "spark-job.py")

            return DataprocJob(
                project="my-project",
                region="us-central1",
                cluster_name="my-cluster",
                job_id=job_id
            )

    Attributes:
        project: GCP project ID
        region: Dataproc region
        cluster_name: Cluster name
        job_id: Job ID
    """

    project: str
    region: str
    cluster_name: str
    job_id: str

    def service_type(self) -> str:
        return "dataproc"

    def tracking_id(self) -> str:
        return f"{self.project}/{self.region}/{self.cluster_name}/{self.job_id}"


@dataclass
class CustomHttp(Deferrable):
    """Custom HTTP polling deferrable

    Used for custom external services that provide HTTP status endpoints.
    The scheduler will poll the URL until the response indicates completion.

    Example:
        from ork_sdk import CustomHttp

        def trigger_external_job():
            job_id = start_external_job()

            return CustomHttp(
                url=f"https://api.example.com/jobs/{job_id}/status",
                method="GET",
                headers={"Authorization": "Bearer token"},
                success_status_codes=[200],
                completion_field="status",
                completion_value="completed",
                failure_value="failed"
            )

    Attributes:
        url: URL to poll for job status
        method: HTTP method (GET, POST, etc.)
        headers: Optional HTTP headers
        success_status_codes: HTTP status codes that indicate successful polling
        completion_field: JSON field to check for completion status
        completion_value: Value indicating job completion
        failure_value: Optional value indicating job failure
    """

    url: str
    method: str = "GET"
    headers: Dict[str, str] = field(default_factory=dict)
    success_status_codes: List[int] = field(default_factory=lambda: [200])
    completion_field: str = "status"
    completion_value: str = "completed"
    failure_value: Optional[str] = None

    def service_type(self) -> str:
        return "custom_http"

    def to_dict(self) -> Dict[str, Any]:
        payload = super().to_dict()
        payload["status_field"] = payload.pop("completion_field")
        payload["success_value"] = payload.pop("completion_value")
        return payload

    def tracking_id(self) -> str:
        # Use URL as job identifier
        return self.url


# Type alias for task results
TaskResult = Union[Any, Deferrable, List[Deferrable]]
